schema-engine/connectors/schema-connector/src/warnings.rs

```diff
@@ -1,130 +1,55 @@
     }
 }
 
-fn fmt_group_view_and_field(f: &mut fmt::Formatter<'_>, view: &String, vec: &[&ViewAndField]) -> fmt::Result {
-    write!(f, r#"View: "{}""#, &view)?;
-    write!(f, ", field(s): [")?;
-
-    let (last, vec_but_last) = vec.split_last().unwrap();
-    for entry in vec_but_last {
-        write!(f, r#""{}", "#, entry.field)?;
-    }
-
-    writeln!(f, r#""{}"]"#, last.field)?;
-
-    Ok(())
-}
-
 impl fmt::Display for GroupBy<'_, ViewAndField> {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        let grouped = self.group_by();
-        for (key, value) in grouped {
-            write!(f, "  - ")?;
-            fmt_group_view_and_field(f, &key, &value)?;
-        }
-
-        Ok(())
+        display_list(self.0, "View", |vf| &vf.view, |vf| &vf.field, f)
     }
 }
 
-impl Groupable<String, TypeAndField> for GroupBy<'_, TypeAndField> {
-    fn group_by(&self) -> BTreeMap<String, Vec<&TypeAndField>> {
-        let mut result: BTreeMap<String, Vec<&TypeAndField>> = BTreeMap::new();
-
-        for item in self.0 {
-            result.entry(item.composite_type.clone()).or_default().push(item);
-        }
-
-        result
-    }
-}
-
-fn fmt_group_type_and_field(f: &mut fmt::Formatter<'_>, r#type: &String, vec: &[&TypeAndField]) -> fmt::Result {
-    write!(f, r#"Composite type: "{}""#, &r#type)?;
-    write!(f, ", field(s): [")?;
-
-    let (last, vec_but_last) = vec.split_last().unwrap();
-    for entry in vec_but_last {
-        write!(f, r#""{}", "#, entry.field)?;
+impl fmt::Display for GroupBy<'_, TypeAndField> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        display_list(self.0, "Composite type", |cf| &cf.composite_type, |cf| &cf.field, f)
     }
-
-    writeln!(f, r#""{}"]"#, last.field)?;
-
-    Ok(())
 }
 
-impl fmt::Display for GroupBy<'_, TypeAndField> {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        let grouped = self.group_by();
-        for (key, value) in grouped {
-            write!(f, "  - ")?;
-            fmt_group_type_and_field(f, &key, &value)?;
+fn display_list<T: Ord>(
+    items: &[T],
+    group_name: &str,
+    project_key: fn(&T) -> &str,
+    project_field: fn(&T) -> &str,
+    f: &mut fmt::Formatter<'_>,
+) -> fmt::Result {
+    let sorted: BTreeSet<_> = items.iter().collect();
+    let mut sorted = sorted.into_iter().peekable();
+    let mut key = None;
+    let close = |f: &mut fmt::Formatter<'_>| f.write_str("]\n");
+
+    while let Some(next) = sorted.next() {
+        if Some(project_key(next)) != key {
+            write!(f, r#"  - {group_name}: "{}", field(s): ["#, project_key(next))?;
+            key = Some(project_key(next));
         }
 
-        Ok(())
+        write!(f, r#""{}""#, project_field(next))?;
+        match sorted.peek() {
+            Some(vf) if Some(project_key(vf)) != key => close(f)?,
+            None => close(f)?,
+            Some(_) => f.write_str(", ")?,
+        }
```

Comment:
Thanks for this snippet of `Iterator::peek()` @tomhoule, now it's much clearer how it should be used.
It reminds me of a `std::next(iter)` wrapped in a `std::optional` in C++17